### PR TITLE
Fixes #13571 - Remote execution integration

### PR DESCRIPTION
**Pull request by iNecas**
Add ability to use the Remote Execution tooling for the content actions
on hosts.

First, we need to use the developer API to define the remote execution features
and templates.

Then, there is ability in UI to jump into the remote execution UI
from the appropriate places.

katello-rex-1

katello-rex-2

Depends on https://github.com/theforeman/foreman_remote_execution/pull/149.

- [x] - detect if Remote Execution plugin is installed and make the option available only when presented
- [x] - errata support

### CODE REVIEW

**Comment by jlsherrill** [comment]
I'm okay if we assume remote execution is installed (and start requiring it by default)

we're going to do that eventually to support things like errata application so might as well do that now?

**Comment by jlsherrill** [comment]
Also, curious if you think it makes sense to be able to specify a default value for whether you want to use remote execution or not. If someone really likes remote execution they may want to do that.

**Comment by jlsherrill** [comment]
[test]

**Comment by ehelms** [comment]
When viewing the gif, what I missed was how a user does a package install as we do them today with gofer/katello-agent. I saw the ability to choose how to perform but both looked via remote execution only.

**Comment by ehelms** [comment]
I would prefer we not assume remote execution is installed and either make it a hard dependency or not. I can see value in both cases:

 1) Hard dependency on remote execution for Katello to work
 2) Optional dependency that unlocks features when present 

Option 2 provides more composibility and lighter-weightness to Katello as it can manage your content but can be enhanced by adding remote execution to unlock functionality.

File: app/controllers/katello/remote_execution_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+module Katello
+  class RemoteExecutionController < JobInvocationsController
```
Comment (ehelms) [inline app/controllers/katello/remote_execution_controller.rb]: Which screen in the diff is this being used at?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js [hunk 2]
```diff
@@ -84,5 +84,22 @@ angular.module('Bastion.content-hosts').controller('ContentHostsBulkActionPackag
             return deferred.promise;
         };
 
+        $scope.performViaRemoteExecution = function(action, customize) {
+            var remoteAction, selectedHosts = $scope.nutupane.getAllSelectedResults();
+            if ($scope.content.contentType == "package_group") {
+                remoteAction = "group_" + action;
+            } else if ($scope.content.contentType == "package") {
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: s/==/===

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js [hunk 3]
```diff
@@ -84,5 +84,22 @@ angular.module('Bastion.content-hosts').controller('ContentHostsBulkActionPackag
             return deferred.promise;
         };
 
+        $scope.performViaRemoteExecution = function(action, customize) {
+            var remoteAction, selectedHosts = $scope.nutupane.getAllSelectedResults();
+            if ($scope.content.contentType == "package_group") {
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: s/==/===

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js [hunk 4]
```diff
@@ -84,5 +84,22 @@ angular.module('Bastion.content-hosts').controller('ContentHostsBulkActionPackag
...
+            }
+            form = $('#packageActionForm');
+            form.attr('action', '/katello/remote_execution');
+            form.attr('method', 'post');
+            form.find('input[name=remote_action]').val(remoteAction);
+            form.find('input[name=authenticity_token]').val(AUTH_TOKEN.replace(/&quot;/g,''));
+            form.find('input[name=customize]').val(customize);
+            form.find('input[name=content_host_ids]').val(selectedHosts.included.ids.join(','));
+            form.find('input[name=scoped_search]').val(selectedHosts.included.search);
+            form.submit();
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: This looks like jquery style JS being put into an Angular controller which we should be able to avoid.

File: lib/katello/plugin.rb [hunk 5]
```diff
@@ -171,4 +171,11 @@
                    :partial => 'smart_proxies/show/storage',
                    :onlyif => proc { |proxy| proxy.has_feature?(SmartProxy::PULP_FEATURE) || proxy.has_feature?(SmartProxy::PULP_NODE_FEATURE) }
   end
+
+  RemoteExecutionFeature.register(:katello_package_install, N_("Katello: Install Package"), :description => N_("Install package via Katello interface"), :provided_inputs => ['package'])
```
Comment (ehelms) [inline lib/katello/plugin.rb]: Do these need to be prepended with 'Katello' ? I am not seeing the value. Can you explain what this means by installing via the Katello interface?

File: app/controllers/katello/remote_execution_controller.rb (hunk 1 above)
Comment (ehelms) [inline app/controllers/katello/remote_execution_controller.rb]: Looking further down the diff, it appears this is providing a controller to hit remotely but not any UI components itself?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js [hunk 6]
```diff
@@ -84,5 +84,22 @@ angular.module('Bastion.content-hosts').controller('ContentHostsBulkActionPackag
...
+        $scope.performViaRemoteExecution = function(action, customize) {
+            var remoteAction, selectedHosts = $scope.nutupane.getAllSelectedResults();
+            if ($scope.content.contentType == "package_group") {
+                remoteAction = "group_" + action;
+            } else if ($scope.content.contentType == "package") {
+                remoteAction = "package_" + action;
+            }
+            form = $('#packageActionForm');
+            form.attr('action', '/katello/remote_execution');
+            form.attr('method', 'post');
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: IF we need to remotely hit an action, can't we build an API around this? That way, like our other APIs, users get new UI and new API and everybody wins.

File: lib/katello/plugin.rb (hunk 5 above)
Comment (iNecas) [inline lib/katello/plugin.rb]: The feature represents a something that some part of code provides. It's declaring, that some part of the system is going to use remote execution for something. I've used the katello prefix, becuase it's really a Katello owning the feature.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 6 above)
Comment (iNecas) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: There is an API in remote execution, that can already be used to achieve this. In fact, I've tried to go the API way first, but failed miserably: my problem was I needed to get to the remote execution UI (either the job invocaiton show page, or the new job invocation form in case of customize first case): I really want to avoid doing the same remote execution UI twice with and without angular. When using the API approach, I then had to use `$window.location.href` for getting to other page + and just translate the API reponse to the `path` to pass to `$window.location.href`. So I explored this path, but turned out to be too complex for what we want to achiieve. Long story short, I ended up going this way because:

1. the main goal for this is getting to the foreman remote execution UI, and I have not found any nice Angular way to do the stuff.
2. there is already API and CLI on the remote execution that can be just used for doing the same thing. Therefore, the added value of having the API on Katello side as well didn't look that interesting.

File: lib/katello/plugin.rb (hunk 5 above)
Comment (ehelms) [inline lib/katello/plugin.rb]: What makes, for example, package installation specific to Katello? I realize we have UI and other precedence for it, but if I am using say Foreman, Foreman Docker and Remote Execution -- why can't I go use a package install feature on those hosts to install a package I need?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 4 above)
Comment (iNecas) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: I'm not saying I'm particularly happy about this approach, but my other tries ended up miserably :/ I also don't have resources to rewrite the whole remote execution interface in Angular and don't think it would move us forward (we would still be a step behind the remote execution UI interface). Trying to do it the angular way just seemed over-complicated, and not worth, since the goal here really is get into the non-angular part of Remote Execution. The last thing that made be use the jQuery was not ability to generate the `get` query from params in pure angular, so I needed to use `$.params` instead, which just made me think that in this particular case, this is the simplest solution.

I'm open to other approaches, but to be honest, I've spent a lot of time trying to go the Angular way, but it just didn't feel right. And if I were you right now seeing the code, I would suggest the same as you do right now:) But with my history of trying, I'm out of other nicer ideas and this brings me just what I need. I realize this is not what Angular people do, but they don't have a hybrid UI to deal with, as we do. 

What I had before was:
1. in the `customize first` case, I needed to construct a get request to `katello/remote_execution/new` page, and use `$window.location.href` to get there
2. in the `run immediately` case, I needed to define a resource in angular, to trigger an API that would return a job invocation id, just to set the `$window.location.href` to `/job_invocations/:id`. And if there was an error while preparing the job invocation, I would need to format that error messages as well.

When going the `jQuery` way, I've solved both cases with one code and, in case of `run immediately` case, I get to the job invocation form with validation errors for free. 

I hope that explains more my intentions and reasons why ending up in this.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 6 above)
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: If there is an API in REX, why can't we form the proper JSON and POST/PUT to it to perform the action? That's the aspect of this that I am missing not knowing the REX plugin.

File: lib/katello/plugin.rb (hunk 5 above)
Comment (iNecas) [inline lib/katello/plugin.rb]: You can. This mostly serves as a mapping to the other templates. This way we make sure, that the default teplates match the inputs with what Katello provides

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 6 above)
Comment (iNecas) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: Because I want to get to the remote execution UI after using the UI feature. Using the API and then using `$window.location.href` to getting to the remote execution screen, is just a more complicated way of doing so as we do right now (+ we loose the ability to show the form with invalid data, where there are some)

**Comment by iNecas** [comment]
@ehelms @jlsherrill I have both:
    - detect if Remote Execution plugin is installed and make the option available only when presented
    - errata support
as a todo in this PR, just wanted to get the initial code into review earlier. So the plan is optional dependency and unlocking the new features when rex is available. The errata support should arrive today. 

The original use for katello-agent is still triggered by the original buttons itselfs. I can make that as another option in the poppup, and make the default configurable in settings: seems like something people, that prefer remote execution would prefer anyway.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 3 above)
Comment (iNecas) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: fixed

File: app/controllers/katello/remote_execution_controller.rb (hunk 1 above)
Comment (iNecas) [inline app/controllers/katello/remote_execution_controller.rb]: It mostly uses the present UI in JobInvocationsController (from remote execution). This provides the mapping between the Katello UI and the remote execution

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 2 above)
Comment (iNecas) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: fixed

**Comment by iNecas** [comment]
I've made the remote execution optiona + a setting option to make it the default action. The popup now contains "run via Katello agent" option as well

**Comment by ehelms** [comment]
@jlsherrill currently this defaults to and shows Katello Agent as the top choice no matter if katello agent is installed - I lean towards not showing that as an option or simply not defaulting to it if not present. What do you think? 

@iNecas consider me a REX noob here, I selected install via ssh and was taken to a job invocation screen where it failed, but I can't tell why it failed. 

Question -- does the SSH option require the content host is registered against a particular proxy or the main proxy with the SSH feature enabled on it?  If yes, should that option be displayed if the content host is or the main proxy doesn't have that feature?

screenshot from 2016-02-09 11-40-09

**Comment by iNecas** [comment]
@ehelms the failure is due to host coming from Katello does not have a operating system assigned. I would like to see that fixed in the Katelo (I can do that) if you think it's a good idea? We could fallback in our template to handle this case though, @stbenjam ideas?

**Comment by iNecas** [comment]
@ehelms ad. the proxy selection, see http://theforeman.org/plugins/foreman_remote_execution/0.2/#4.1Determiningthesmartproxyforhost : there are multiple strategies for selecting one

**Comment by stbenjam** [comment]
> @ehelms the failure is due to host coming from Katello does not have a operating system assigned. I would like to see that fixed in the Katelo (I can do that) if you think it's a good idea? We could fallback in our template to handle this case though, @stbenjam ideas?

In a real world situation Katello hosts created through subscription-manager will have an Operating System since https://github.com/Katello/katello/pull/5719.  I think it'd be better if the test objects reflected that.

It raises an issue though that the template doesn't support anything other than Red Hat and Debian, so we should handle that case -  by like as you suggested setting a default (e.g. yum), or failing better.  I think I'd prefer the latter, because someone should running Package Action on FreeBSD should get some helpful message.

**Comment by stbenjam** [comment]
Oh I see, not a test object but that was @ehelms's real host - how come it doesn't have an OS assigned? Both puppet reported hosts and subscription-manager reported hosts should have one.

**Comment by ehelms** [comment]
Is this rebased to include #5719? I am not seeing in the git logs and I registered this host against this PR.

**Comment by stbenjam** [comment]
Yea, it's not, it was a recent merge.  Rebasing should solve the problem. 

 The error message should be more helpful anyway, I filed http://projects.theforeman.org/issues/13632.

**Comment by iNecas** [comment]
I've rebased the stuff and converted the `content_host_ids`

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js (hunk 6 above)
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/bulk/content-hosts-bulk-action-packages.controller.js]: I created a patch for this that uses the angularJS style of doing things:

https://github.com/iNecas/katello/pull/5

**Comment by iNecas** [comment]
Ad. not showing the stuff when the rex/agent is not available for the host: it's a bit tricky, especially for the bulk action, where we would need to check if at least on from the group has the feature. Also, in case you want to plan the execution in the future, in theory, you would not need the host to have the proxy available right now. I would rather see this handled in remote execution by enhancing the error messages, when no proxy is available for the execution

**Comment by iNecas** [comment]
I've taken @waldenraines suggestion and improved the JS code, it looks a bit more as Angular code now. The CI should get green now.

**Comment by iNecas** [comment]
Tests are green now

**Comment by ehelms** [comment]
I will take a look at this again today - thanks for indulging us with getting the UI code up to speed!

**Comment by iNecas** [comment]
I've done some updates to the developer api on rex side.  If you rebase https://github.com/theforeman/foreman_remote_execution/pull/149 on already migrated system, you should follow https://github.com/theforeman/foreman_remote_execution/pull/149 to get up to date with the schema (or just resetting the databases form scratch).

**Comment by ehelms** [comment]
screenshot from 2016-02-15 13-18-18

I see the above error when templates aren't associated which is kind of ugly from a user perspective. Further, after a complete reset with migrate and seed I do not see the templates auto-associated to the various features. Is that intentional? I thought when I previously reviewed it they were automatically associated.

**Comment by iNecas** [comment]
It gets associated only at a time of first seeding the templates: the reason for this is we don't know, if no template is assigned to the template on purpose or accidentally.

We could say though that we don't allow no template to be assigned to the feature, in which case we would assign it to the seeded one when seeding.

**Comment by ares** [comment]
after short discussion with Ivan, I lean to that we should re-associate the template as the on purpose disassociation is quite edge case and I'd recommend using permissions for this use case

**Comment by ehelms** [comment]
1) So don't allow it to be unset? or default to the base if not otherwise specified? I just want to clarify what we are landing on so I can properly review the behavior.

2) Are the test failures due to needing a new release of foreman_tasks?

**Comment by iNecas** [comment]
1) the desired behaviour will be, if unset, re-attach the original template when seeding, /me waiting for https://github.com/theforeman/foreman_remote_execution/pull/150 to finish tests so that I can push my changes in
2) the issues were fixed by https://github.com/theforeman/foreman-tasks/pull/173#issuecomment-184644119, once I update this PR with latest changes, it should get green

**Comment by ehelms** [comment]
Ok - can a template be complete unset from a feature? In other words, can the template for a feature be set to blank by the user?

**Comment by iNecas** [comment]
yes

**Comment by iNecas** [comment]
I've updated from latest PR in remote execution https://github.com/theforeman/foreman_remote_execution/pull/149: now the unassigned feature should get assigned to the default template when re-seeding. Still I would consider this as a corner case. The tests should get green as well now.

**Comment by iNecas** [comment]
The tests are green :) http://ci.theforeman.org/job/test_katello_pull_request/8431/

**Comment by ehelms** [comment]
ACK - thanks @iNecas !
